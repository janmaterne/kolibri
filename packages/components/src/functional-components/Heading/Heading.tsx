import { h, type FunctionalComponent as FC } from '@stencil/core';
import clsx from 'clsx';
import type { JSXBase, VNode } from '@stencil/core/internal';
import type { HeadingLevel, HeadingVariantPropType } from '../../schema';

type LevelProps = { level?: HeadingLevel };
type VariantProps = { variant?: HeadingVariantPropType };

type BaseProps = JSXBase.HTMLAttributes<HTMLHeadingElement> & LevelProps;
type HeadlineProps = BaseProps & VariantProps;
type SecondaryHeadlineProps = BaseProps;

export type HeadingProps = BaseProps &
	VariantProps & {
		secondaryHeadline?: string;
	};

function getHeadlineTag(level: HeadingLevel | number): string {
	const validHeadline = level > 0 && level <= 6;
	return validHeadline ? `h${level}` : 'strong';
}

function getSubHeadlineTag(level: HeadingLevel | number): string {
	if (level === 1) {
		return 'span';
	}

	return getHeadlineTag(level);
}

const KolHeadlineFc: FC<HeadlineProps> = (props, children) => {
	const { class: classNames, level = 1, variant: defaultVariant, ...other } = props;

	const HeadlineTag = getHeadlineTag(level);
	const variant = defaultVariant || HeadlineTag;

	return (
		<HeadlineTag class={clsx('headline', `headline-${variant}`, classNames)} {...other}>
			{children}
		</HeadlineTag>
	);
};

const KolSecondaryHeadlineFc: FC<SecondaryHeadlineProps> = (props, children) => {
	const { class: classNames, level = 1, ...other } = props;

	const HeadlineTag = getSubHeadlineTag(level + 1);

	return (
		<HeadlineTag class={clsx('secondary-headline', classNames)} {...other}>
			{children}
		</HeadlineTag>
	);
};

const KolHeadingFc: FC<HeadingProps> = (props, children) => {
	const { ref, secondaryHeadline, level = 1, class: classNames, ...other } = props;

	let result: unknown = (
		<KolHeadlineFc ref={ref} level={level} {...other}>
			{children}
		</KolHeadlineFc>
	);

	if (secondaryHeadline) {
		result = (
			<hgroup>
				{result}
				<KolSecondaryHeadlineFc level={level}>{secondaryHeadline}</KolSecondaryHeadlineFc>
			</hgroup>
		);
	}

	return <section class={clsx('kol-heading-wc', classNames)}>{result as VNode}</section>;
};

export default KolHeadingFc;
