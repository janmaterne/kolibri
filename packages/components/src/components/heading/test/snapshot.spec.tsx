import { KolHeadingTag } from '@component-names';
import type { HeadingProps } from '@schema';
import { executeSnapshotTests } from '@testing';

import { KolHeading } from '../Heading';

executeSnapshotTests<HeadingProps>(
	KolHeadingTag,
	[KolHeading],
	[
		{ _label: 'Headline' },
		...[0, 1, 2, 3, 4, 5, 6].map((_level) => ({ _label: 'Headline', _level }) as HeadingProps),
		...['h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'strong'].map((_variant) => ({ _label: 'Headline', _variant }) as HeadingProps),
	],
);
