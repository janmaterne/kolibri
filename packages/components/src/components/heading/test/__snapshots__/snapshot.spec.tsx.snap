// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kol-heading should render with _label="Headline" _level=0 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <strong class="headline headline-strong">
        Headline
        <slot name="expert" slot="expert"></slot>
      </strong>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _level=1 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h1">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _level=2 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h2 class="headline headline-h2">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h2>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _level=3 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h3 class="headline headline-h3">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h3>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _level=4 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h4 class="headline headline-h4">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h4>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _level=5 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h5 class="headline headline-h5">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h5>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _level=6 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h6 class="headline headline-h6">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h6>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="h1" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h1">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="h2" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h2">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="h3" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h3">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="h4" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h4">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="h5" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h5">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="h6" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h6">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" _variant="strong" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-strong">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;

exports[`kol-heading should render with _label="Headline" 1`] = `
<kol-heading class="kol-heading">
  <template shadowrootmode="open">
    <section class="kol-heading-wc">
      <h1 class="headline headline-h1">
        Headline
        <slot name="expert" slot="expert"></slot>
      </h1>
    </section>
  </template>
</kol-heading>
`;
